src/Microsoft.AspNetCore.Mvc.Razor/Internal/CompilerCache.cs

```diff
@@ -126,19 +136,19 @@ public CompilerCache(IFileProvider fileProvider)
 
                 cacheEntryOptions = new MemoryCacheEntryOptions();
 
-                compilerCacheContext = cacheContextFactory(normalizedPath);
-                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(compilerCacheContext.ProjectItem.Path));
-                if (!compilerCacheContext.ProjectItem.Exists)
+                cacheEntryOptions.ExpirationTokens.Add(_fileProvider.Watch(normalizedPath));
+                var projectItem = _templateEngine.Project.GetItem(normalizedPath);
```

Comment:
We need some API to give us the import hierarchy \ import file name. Nothing outside of the template engine has those bits of data today.